### PR TITLE
Buffered reverse-complement studies

### PR DESCRIPTION
**Pull request by benharsh**
These versions are based off of Rust entries 2 and 3, and mimic Rust's ``BufReader`` struct. One version reads entire sections at a time, and the other reads line-by-line. Both versions use a buffer to reduce the number of IO operations, and copy bytes into a given 1D array up until a particular character (> or \n) is found.

The rust versions parallelize differently, but these entries are mainly to test an alternate approach to IO.

The line-by-line version is currently performing significantly worse, likely due to slicing overhead. PR #8022 added a test dedicated to measuring creation time for array views.

The 'entire sections at a time' version is competitive with the top C entry.

### CODE REVIEW

**Comment by benharsh** [comment]
@chapel-lang/perf-team : anyone want to review?

File: test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  proc init(fi:file, bs:int) {
+    this.bufSize = bs;
+
+    super.init();
+
+    chan = fi.reader(locking=false);
+    numLeft = fi.length();
+  }
+
+  pragma "no copy return"
```
Comment (mppf) [inline test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl]: Can you put a comment saying that this returns a view into the buffer starting at low ?

File: test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  proc readUntil(term : uint(8), data : [] uint(8)) : int {
+    var read = 0;
+    while true {
+      var done = false, used = 0;
+      ref avail = fill();
+      if avail.size > 0 {
+        const idx = _memchr(term, avail);
+        if idx >= 0 {
+          data.push_back(avail[..idx]);
```
Comment (mppf) [inline test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl]: For a while I was confused about this line. Could you add a comment here or at the function level, to indicate that it appends a big block to the array once it's available?

File: test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    return 0;
+  }
+}
+
+config const readSize = 16 * 1024;
+
+proc main(args: [] string) {
+  const stdin = openfd(0);
+  var input = new buf(stdin, readSize);
+  var data : [1..0] uint(8);
```
Comment (mppf) [inline test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl]: It'd help to have some comments here. For one thing, you're trying to make `data` have enough storage to store all of the bytes in the file.

**Pull request merged**